File: app/src/main/java/org/thoughtcrime/securesms/service/NotificationController.java
package org.thoughtcrime.securesms.service;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;

import androidx.annotation.NonNull;

import java.util.concurrent.atomic.AtomicReference;

public final class NotificationController implements AutoCloseable {

  private final @NonNull Context context;
  private final int id;

  private int     progress;
  private int     progressMax;
  private boolean indeterminate;
  private long    percent = -1;

  private final AtomicReference<GenericForegroundService> service = new AtomicReference<>();

  NotificationController(@NonNull Context context, int id) {
    this.context = context;
    this.id      = id;

    bindToService();
  }

  private void bindToService() {
    context.bindService(new Intent(context, GenericForegroundService.class), new ServiceConnection() {
      @Override
      public void onServiceConnected(ComponentName name, IBinder service) {
        GenericForegroundService.LocalBinder binder       = (GenericForegroundService.LocalBinder) service;
        GenericForegroundService genericForegroundService = binder.getService();

        NotificationController.this.service.set(genericForegroundService);

        updateProgressOnService();
      }

      @Override
      public void onServiceDisconnected(ComponentName name) {
        service.set(null);
      }
    }, Context.BIND_AUTO_CREATE);
  }

  public int getId() {
    return id;
  }

  @Override
  public void close() {
    GenericForegroundService.stopForegroundTask(context, id);
  }

  public void setIndeterminateProgress() {
    setProgress(0, 0, true);
  }

  public void setProgress(long newProgressMax, long newProgress) {
    setProgress((int) newProgressMax, (int) newProgress, false);
  }

  private synchronized void setProgress(int newProgressMax, int newProgress, boolean indeterminant) {
    int newPercent = newProgressMax != 0 ? 100 * newProgress / newProgressMax : -1;

    boolean same = newPercent == percent && indeterminate == indeterminant;

    percent       = newPercent;
    progress      = newProgress;
    progressMax   = newProgressMax;
    indeterminate = indeterminant;

    if (same) return;

    updateProgressOnService();
  }

  private synchronized void updateProgressOnService() {
    GenericForegroundService genericForegroundService = service.get();

    if (genericForegroundService == null) return;

    genericForegroundService.replaceProgress(id, progressMax, progress, indeterminate);
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/service/NotificationController.kt
package org.thoughtcrime.securesms.service

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.IBinder
import org.signal.core.util.logging.Log
import org.thoughtcrime.securesms.jobs.UnableToStartException
import org.thoughtcrime.securesms.service.GenericForegroundService.Companion.stopForegroundTask
import org.thoughtcrime.securesms.service.GenericForegroundService.LocalBinder
import java.util.concurrent.atomic.AtomicReference
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

class NotificationController internal constructor(private val context: Context, val id: Int) : AutoCloseable, ServiceConnection {
  private val service = AtomicReference<GenericForegroundService?>()
  private val lock = ReentrantLock()

  private var progress = 0
  private var progressMax = 0
  private var indeterminate = false
  private var percent: Int = -1
  private var isBound: Boolean

  companion object {
    private val TAG = Log.tag(NotificationController::class.java)
  }

  init {
    isBound = bindToService()
  }

  override fun onServiceConnected(name: ComponentName, service: IBinder) {
    Log.i(TAG, "[onServiceConnected] Name: $name")

    val binder = service as LocalBinder
    val genericForegroundService = binder.service

    this.service.set(genericForegroundService)

    lock.withLock {
      updateProgressOnService()
    }
  }

  override fun onServiceDisconnected(name: ComponentName) {
    Log.i(TAG, "[onServiceDisconnected] Name: $name")
    service.set(null)
  }

  override fun close() {
    try {
      if (isBound) {
        Log.d(TAG, "[close] Unbinding service.")
        context.unbindService(this)
        isBound = false
      } else {
        Log.w(TAG, "[close] Service was not bound at the time of close()...")
      }

      if (service.get()?.hasTimedOut == true) {
        Log.w(TAG, "[close] Service has timed out, skipping stop foreground task.")
      } else {
        stopForegroundTask(context, id)
      }
    } catch (e: IllegalStateException) {
      Log.w(TAG, "[close] Failed to unbind service...", e)
    } catch (e: UnableToStartException) {
      Log.w(TAG, "[close] Failed to unbind service...", e)
    }
  }

  fun setIndeterminateProgress() {
    lock.withLock {
      setProgress(
        newProgressMax = 0,
        newProgress = 0,
        indeterminant = true
      )
    }
  }

  fun setProgress(newProgressMax: Long, newProgress: Long) {
    lock.withLock {
      setProgress(
        newProgressMax = newProgressMax.toInt(),
        newProgress = newProgress.toInt(),
        indeterminant = false
      )
    }
  }

  fun replaceTitle(title: String) {
    lock.withLock {
      service.get()?.replaceTitle(id, title)
        ?: Log.w(TAG, "Tried to update the title, but the service was no longer bound!")
    }
  }

  private fun bindToService(): Boolean {
    return context.bindService(Intent(context, GenericForegroundService::class.java), this, Context.BIND_AUTO_CREATE)
  }

  private fun setProgress(newProgressMax: Int, newProgress: Int, indeterminant: Boolean) {
    val newPercent = if (newProgressMax != 0) {
      100 * newProgress / newProgressMax
    } else {
      -1
    }

    val same = newPercent == percent && indeterminate == indeterminant

    percent = newPercent
    progress = newProgress
    progressMax = newProgressMax
    indeterminate = indeterminant

    if (!same) {
      updateProgressOnService()
    }
  }

  private fun updateProgressOnService() {
    service.get()?.replaceProgress(id, progressMax, progress, indeterminate)
      ?: Log.w(TAG, "Tried to update the progress, but the service was no longer bound!")
  }
}
